packages/xstate-solid/src/deriveServiceState.ts

```diff
@@ -0,0 +1,64 @@
+import type {
+  ActorRef,
+  Interpreter,
+  InterpreterFrom,
+  StateLike,
+  State
+} from 'xstate';
+
+export function isStateLike(state: any): state is StateLike<any> {
+  return (
+    typeof state === 'object' &&
+    'value' in state &&
+    'context' in state &&
+    'event' in state &&
+    '_event' in state
+  );
+}
+
+function isService(value: any): value is Interpreter<any> {
+  return 'state' in value && 'machine' in value;
+}
+
+/**
+ * Takes in an interpreter or actor ref and returns a State object with reactive
+ * methods or if not State, the initial value passed in
+ * @param service {InterpreterFrom<any> | ActorRef<any>}
+ * @param initialState {State<any> | unknown}
+ */
+export const deriveServiceState = <
+  TService extends InterpreterFrom<any> | ActorRef<any>,
+  InitialState extends State<any>
+>(
+  service: TService,
+  initialState: InitialState extends unknown ? unknown : InitialState
+): InitialState => {
+  if (isService(service) && isStateLike(initialState)) {
+    return {
+      ...(initialState as object),
+      toJSON() {
+        return service.state.toJSON();
+      },
+      toStrings(...args: Parameters<InitialState['toStrings']>) {
+        return service.state.toStrings(args[0], args[1]);
+      },
+      can(...args: Parameters<InitialState['can']>) {
+        // tslint:disable-next-line:no-unused-expression
+        this.value; // reads state.value to be tracked
```

Comment:
This looks incorrect because `state.can` is not exactly directly related to `state.value`. For two states with the same value, it could return a different result. At the very least I would also "subscribe" here to `state.context`

Reply:
could we also add a test case that would cover this change?